File: lib/models/product.dart
import 'dart:convert';

import 'package:my_amazon/models/rating.dart';

class Product {
  final String name;
  final String description;
  final double quantity;
  final List<String> images;
  final String category;
  final double price;
  final String? id;
  final List<Rating>? ratings;

  Product(
      {required this.name,
      required this.description,
      required this.quantity,
      required this.images,
      required this.category,
      required this.price,
      this.id,
      this.ratings});

  factory Product.fromJson(Map<String, dynamic> map) {
    return Product(
      id: map['_id'] ?? '',
      name: map['name'] ?? '',
      description: map['description'] ?? '',
      quantity: map['quantity']?.toDouble() ?? 0,
      price: map['price'].toDouble() ?? 0,
      images: List<String>.from(map['images']),
      category: map['category'] ?? '',
      ratings: map['ratings'] != null ? List<Rating>.from(map['ratings']?.map((e)=> Rating.fromMap(e))):null
    );
  }

  Map<String, dynamic> toMap() {
    return {
      "_id": id,
      "name": name,
      "description": description,
      "images": images,
      "quantity": quantity,
      "price": price,
      "category": category,
    };
  }

  String toJson() => jsonEncode(toMap());
}

File: server/models/product.js
const mongoose = require('mongoose');
const ratingSchema = require('./rating');

const productSchema = mongoose.Schema(
    {
        name:{
            required:true,
            type:String,
            trim:true
        },
        description:{
            required:true,
            type:String,
            trim:true
        },
        images:[
            {
                required:true,
                type:String,
            }
        ],
        quantity:{
            type:Number,
            required:true
        },
        price:{
            type:Number,
            required:true
        },
        category:{
            type:String,
            required:true
        },
        ratings:[ratingSchema]
    }
);

const Product = mongoose.model("Product",productSchema);

module.exports = {Product, productSchema};